### PR TITLE
useExportData

### PR DESCRIPTION
**Pull request by gargroh**
Added new react-table hook for exporting data. Currently it supports csv export, but other export can be added easily in it.

It support :- 
1. Exporting current view
2. Export all data
3. Overriding exported file name
4. Overriding Column header name
5. Overriding Cell value
6. Disable export at column and table level
7. Multiple file type exports(currently csv is implemented)

Few areas where @tannerlinsley your thoughts/help may be required :- 
1. I have not considered pagination support.
2. How it should behave for grouped/hierarchical data
3. Web workers support.
4. Can we avoid importing third party libs in hooks(e.g. for csv i used papaparse here)
5. Testing

### CODE REVIEW

**Comment by ee0pdt** [comment]
> 4. Can we avoid importing third party libs in hooks(e.g. for csv i used papaparse here)

One option here would be to change the API to permit sending a parser to the hook,

e.g. Something like

```
  const csvParser = (data) => {
    const csvString = Papa.unparse(data)
    return new Blob([csvString], { type: 'text/csv' })
  }

...

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    exportData,
  } = useTable(
    {
      columns,
      data,
      defaultColumn,
    },
    useFilters,
    useSortBy,
    useExportData({parser: csvParser, fileExtension: 'csv')
  )
```

**Comment by tannerlinsley** [comment]
> Added new react-table hook for exporting data. Currently it supports csv export, but other export can be added easily in it.
> 
> It support :-
> 
> 1. Exporting current view
> 2. Export all data
> 3. Overriding exported file name
> 4. Overriding Column header name
> 5. Overriding Cell value
> 6. Disable export at column and table level
> 7. Multiple file type exports(currently csv is implemented)
> 
> Few areas where @tannerlinsley your thoughts/help may be required :-
> 
> 1. I have not considered pagination support.

As long as it uses `rows` and not `page` for getting the rows to export it should be fine. You can't really do a full export on server-paginated data anyway, but if the pagination is client-side, you likely don't want the page separation anyway.

> 2. How it should behave for grouped/hierarchical data

Easiest solution. Don't support it for now. Otherwise, it could export columns like row ID, depth, path etc to denote this structure

> 3. Web workers support.

This would likely be the parser's job, so I think just providing the right optional hook points to outsource potentially expensive work is fine. Right now all of the calculation is memoized in the table anyway, so it's just transposing the data structure into a string format, right? Doesn't seem that bad right now.

> 4. Can we avoid importing third party libs in hooks(e.g. for csv i used papaparse here)

Like in your comment above, the user should be **required** to supply a stringify function eg. like papaparse. I don't think it should ship with a native one, but instead show a few examples on how to use existing ones like papaparse or excel-js or something else.

> 5. Testing

No ideas here yet.

**Comment by gargroh** [comment]
Hey @tannerlinsley this is ready to review.

File: examples/export-data/.rescriptsrc.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    modules: [path.resolve('src'), ...config.resolve.modules],
+  }
+  return config
+}
+
+module.exports = [
+  ['use-babel-config', '.babelrc'],
+  ['use-eslint-config', '.eslintrc'],
+  fixLinkedDependencies,
+  // includeSrcDirectory,
```
Comment (omridevk) [inline examples/export-data/.rescriptsrc.js:28]: comment

**Comment by omridevk** [comment]
+1 for this feature. very cool PR!

**Comment by GuptaSiddhant** [comment]
I created typescript declaration for this plugin.
```
// useExportData.d.ts
import { Hooks, ColumnInstance, Row } from "react-table";
declare type HeaderName = string | number | object | undefined;
declare type FileType = "csv";
export interface UseExportDataColumnProps<T extends object> {
    canExport: boolean;
    disableExport: boolean;
    getExportHeaderValue: (col: ColumnInstance<T>) => HeaderName;
    getExportCellValue: (row: Row<T>, col: {
        id: string;
        name: HeaderName;
        columnObj: ColumnInstance<T>;
    }) => any;
}
export interface UseExportDataInstanceProps {
    exportData: ({ all, type, fileName, }: {
        all?: boolean;
        type?: FileType;
        fileName?: string;
    }) => void;
}
/**
 * @author Rohit Garg
 * @author Siddhant Gupta
 */
export declare const useExportData: {
    <D extends object = {}>(hooks: Hooks<D>): void;
    pluginName: string;
};
export default useExportData;
```

It extends the react-table-config.d.ts as follows:
```
export interface ColumnInstance<D extends object = {}>
  extends UseExportDataColumnProps<D> {}

export interface TableInstance<D extends object = {}>
    extends UseExportDataInstanceProps {}
```

**Comment by tannerlinsley** [comment]
@gargroh Same for this one. The plugin system is changing a bit, so I'd like to see this ported to v8! This is going to be a great feature.

**Comment by RajuTicvic** [comment]
> > 1. Can we avoid importing third party libs in hooks(e.g. for csv i used papaparse here)
> 
> One option here would be to change the API to permit sending a parser to the hook,
> 
> e.g. Something like
> 
> ```
>   const csvParser = (data) => {
>     const csvString = Papa.unparse(data)
>     return new Blob([csvString], { type: 'text/csv' })
>   }
> 
> ...
> 
>   const {
>     getTableProps,
>     getTableBodyProps,
>     headerGroups,
>     rows,
>     prepareRow,
>     exportData,
>   } = useTable(
>     {
>       columns,
>       data,
>       defaultColumn,
>     },
>     useFilters,
>     useSortBy,
>     useExportData({parser: csvParser, fileExtension: 'csv'})
>   )
> ```
> 
> This would avoid the expense of including a third party library and allow users to easily add their own parsers to .xls and PDF etc.

Is this ready to use
if yes,
can you please share the URL

**Comment by gargroh** [comment]
>@gargroh Same for this one. The plugin system is changing a bit, so I'd like to see this ported to v8! This is going to be a great feature.

@tannerlinsley till the time v8 is launched(which is going to take considerable amount of time) and this is incorporated in it, i have created https://github.com/gargroh/react-table-plugins repository which will hold some of the plugins like  `useExportData` etc. 

In this way nobody will be blocked on any plugin which are not currently present in v7 and you can also focus on v8. Developers can raise PR in react table v7 for any needed properties, hooks etc. if they require to use in these plugins. Like https://github.com/tannerlinsley/react-table/pull/2400, which is required by `useExportData` plugin to get rows before sort, filter etc.

I have also given examples and documentation so as users find it easy to adopt and use them.

Closing this PR.

**Comment by codercodingthecode** [comment]
Having a hard time with this and typescript to work together.

**Comment by Ac-Srikanth** [comment]
@gargroh,  man thank you so much for that plugin ,  Took me like 5 mins to incorporate that . Sandbox example really helped. 
Cheers :)

**Comment by FelipeMartini** [comment]
How can I use this?

**Comment by dawoodjehangir** [comment]
> Having a hard time with this and typescript to work together.
> 
> edit: Got it to work. But doc was tricky.

@codercodingthecode how did you got this to work with typescript?

**Comment by Dev-eloperr** [comment]
@codercodingthecode Can you please share how you got this to work with TS. 
or @dawoodjehangir if you have figured out, can you share please?

**Comment by brianwachira** [comment]
> @codercodingthecode Can you please share how you got this to work with TS. or @dawoodjehangir if you have figured out, can you share please?

@Dev-eloperr @codercodingthecode or @dawoodjehangir  could you share how you got it to work with typescript please?

**Comment by codercodingthecode** [comment]
Sorry, I no longer use this for some time. I currently use mui data grid. I should've shared the solution at the time. During the weekend I will look it up.
@dawoodjehangir @Dev-eloperr @brianwachira

**Comment by Dev-eloperr** [comment]
@brianwachira Hey, Sorry even I forgot how I made it work back then and the exact issue that I was facing. I checked the code and I have this file declared in my types
`filename: useExportData.d.ts`

```
import { Hooks, ColumnInstance, Row } from "react-table";

declare type HeaderName = string | number | object | undefined;
declare type FileType = "csv";

export interface UseExportDataColumnProps<T extends object> {
	canExport: boolean;
	disableExport: boolean;
	getExportHeaderValue: (col: ColumnInstance<T>) => HeaderName;
	getExportCellValue: (
		row: Row<T>,
		col: {
			id: string;
			name: HeaderName;
			columnObj: ColumnInstance<T>;
		}
	) => any;
}
export interface UseExportDataInstanceProps {
	exportData: ({
		all,
		type,
		fileName,
	}: {
		all?: boolean;
		type?: FileType;
		fileName?: string;
	}) => void;
}
/**
 * @author Rohit Garg
 * @author Siddhant Gupta
 */
export declare const useExportData: {
	<D extends object = {}>(hooks: Hooks<D>): void;
	pluginName: string;
};

// export interface ColumnInstance<D extends object = {}>
// extends UseExportDataColumnProps<D> {}

// export interface TableInstance<D extends object = {}>
// extends UseExportDataInstanceProps {}

export default useExportData;

```

**Pull request closed without merging**